tests/unit/lms/services/canvas_api/public_test.py

```diff
@@ -0,0 +1,238 @@
+            [{"id": 1, "name": "name"}, {"id": 1, "name": "DIFFERENT"}]
+        )
+
+        with pytest.raises(CanvasAPIError):
+            canvas_api_client.course_sections("course_id")
+
+    def test_course_sections_raises_CanvasAPIError_with_too_few_returned(
+        self, canvas_api_client, http_session
+    ):
+        http_session.set_response([])
+
+        with pytest.raises(CanvasAPIError):
+            canvas_api_client.course_sections("dummy")
+
+    def test_users_sections(self, canvas_api_client, http_session):
+        http_session.set_response(
+            {
+                "enrollments": [
+                    {"course_section_id": 101, "unexpected": "ignored"},
+                    {"course_section_id": 102, "unexpected": "ignored"},
+                ]
+            }
+        )
+
+        response = canvas_api_client.users_sections("USER_ID", "COURSE_ID")
+
+        assert response == [{"id": 101}, {"id": 102}]
+        http_session.send.assert_called_once_with(
+            Any.request(
+                "GET",
+                url=Any.url.with_path(
+                    "api/v1/courses/COURSE_ID/users/USER_ID"
+                ).with_query({"include[]": "enrollments"}),
+            ),
+            timeout=Any(),
+        )
+
+    def test_users_sections_deduplicates_sections(
+        self, canvas_api_client, http_session
+    ):
+        http_session.set_response(
+            {"enrollments": [{"course_section_id": 1}, {"course_section_id": 1}]}
+        )
+
+        sections = canvas_api_client.users_sections("user_id", "course_id")
+
+        assert sections == [{"id": 1}]
+
+    def test_list_files(self, canvas_api_client, http_session):
+        files = [
+            {"display_name": "display_name_1", "id": 1, "updated_at": "updated_at_1"},
+            {"display_name": "display_name_1", "id": 1, "updated_at": "updated_at_1"},
+        ]
+        files_with_noise = [dict(file, unexpected="ignored") for file in files]
+        http_session.set_response(files_with_noise)
+
+        response = canvas_api_client.list_files("COURSE_ID")
+
+        assert response == files
+
+        http_session.send.assert_called_once_with(
+            Any.request(
+                "GET",
+                url=Any.url.with_path("api/v1/courses/COURSE_ID/files").with_query(
+                    {"content_types[]": "application/pdf", "per_page": Any.string(),}
+                ),
+            ),
+            timeout=Any(),
+        )
+
+    def test_public_url(self, canvas_api_client, http_session):
+        http_session.set_response({"public_url": "public_url_value"})
+
+        response = canvas_api_client.public_url("FILE_ID")
+
+        assert response == "public_url_value"
+        http_session.send.assert_called_once_with(
+            Any.request(
+                "GET", url=Any.url.with_path("api/v1/files/FILE_ID/public_url")
+            ),
+            timeout=Any(),
+        )
+
+
+class TestMetaBehavior:
```

Comment:
As a result of pulling out the big parameterised tests from before there's a lot less in here. These functions test behavior common to many functions.